core/struct.rbs

```diff
@@ -1,299 +1,51 @@
-  #     PositionalOnly.new(0, 1)
-  #     # => #<struct PositionalOnly foo=0, bar=1>
-  #     PositionalOnly.new(bar: 1, foo: 0)
-  #     # => #<struct PositionalOnly foo={:foo=>1, :bar=>2}, bar=nil>
-  #     # Note that no error is raised, but arguments treated as one hash value
-  #
-  #     # Same as not providing keyword_init:
-  #     Any = Struct.new(:foo, :bar, keyword_init: nil)
-  #     Any.new(foo: 1, bar: 2)
-  #     # => #<struct Any foo=1, bar=2>
-  #     Any.new(1, 2)
-  #     # => #<struct Any foo=1, bar=2>
-  #
-  def initialize: (attribute_name, *attribute_name, ?keyword_init: boolish) ?{ () -> void } -> void
+  def self.members: () -> Array[Symbol]
+  def self.keyword_init?: () -> bool?
 
-  # <!--
-  #   rdoc-file=struct.c
-  #   - each {|value| ... } -> self
-  #   - each -> enumerator
-  # -->
-  # Calls the given block with the value of each member; returns `self`:
-  #
-  #     Customer = Struct.new(:name, :address, :zip)
-  #     joe = Customer.new("Joe Smith", "123 Maple, Anytown NC", 12345)
-  #     joe.each {|value| p value }
-  #
-  # Output:
-  #
-  #     "Joe Smith"
-  #     "123 Maple, Anytown NC"
-  #     12345
-  #
-  # Returns an Enumerator if no block is given.
-  #
-  # Related: #each_pair.
-  #
-  def each: () -> ::Enumerator[Elem?, self]
-          | () { (Elem? item) -> void } -> self
+  def ==: (untyped other) -> bool
+  def eql?: (untyped other) -> bool
+  def hash: () -> Integer
 
-  # <!--
-  #   rdoc-file=struct.c
-  #   - StructClass::members -> array_of_symbols
-  # -->
-  # Returns the member names of the Struct descendant as an array:
-  #
-  #     Customer = Struct.new(:name, :address, :zip)
-  #     Customer.members # => [:name, :address, :zip]
-  #
-  def self.members: () -> ::Array[Symbol]
+  def inspect: () -> String
+  alias to_s inspect
 
-  # <!--
-  #   rdoc-file=struct.c
-  #   - StructClass::keyword_init? -> true or falsy value
-  # -->
-  # Returns `true` if the class was initialized with `keyword_init: true`.
-  # Otherwise returns `nil` or `false`.
-  #
-  # Examples:
-  #     Foo = Struct.new(:a)
-  #     Foo.keyword_init? # => nil
-  #     Bar = Struct.new(:a, keyword_init: true)
-  #     Bar.keyword_init? # => true
-  #     Baz = Struct.new(:a, keyword_init: false)
-  #     Baz.keyword_init? # => false
-  #
-  def self.keyword_init?: () -> (true | false | nil)
+  def to_a: () -> Array[Elem?]
+  def to_h: () -> Hash[Symbol, Elem?]
+          | [K, V] () { (Symbol key, Elem? value) -> [K, V] } -> Hash[K, V]
+  alias values to_a
+
+  def size: () -> Integer
+  alias length size
+
+  def each: () -> Enumerator[Elem?, self]
+          | () { (Elem? value) -> void } -> self
+
+  def each_pair: () -> Enumerator[[Symbol, Elem?], self]
+               | () { ([Symbol, Elem?] key_value) -> void } -> self
+
+  def []: (index name_or_position) -> untyped
+  def []=: (index name_or_position, Elem value) -> Elem
+         | (index name_or_position, nil) -> nil
```

Comment:
```suggestion
  def []=: [E < Elem?] (index name_or_position, E) -> E
```
?

and there ought to be many more `[]=`s and `attr_writer`s out there that don’t make use of the type variable

Reply:
That doesn't work, the rhs of `<` needs to be interfaces, otherwise stuff like this would be significantly easier:
```
  # <!--
  #   rdoc-file=vm_method.c
  #   - public                                -> nil
  #   - public(method_name)                   -> method_name
  #   - public(method_name, method_name, ...) -> array
  #   - public(array)                         -> array
  # -->
  # With no arguments, sets the default visibility for subsequently defined
  # methods to public. With arguments, sets the named methods to have public
  # visibility. String arguments are converted to symbols. An Array of Symbols
  # and/or Strings is also accepted. If a single argument is passed, it is
  # returned. If no argument is passed, nil is returned. If multiple arguments are
  # passed, the arguments are returned as an array.
  #
  def public: () -> nil
            | (Symbol method_name) -> Symbol
            | (Symbol, Symbol, *Symbol method_names) -> Array[Symbol]
            | [T < _ToStr] (T method_name) -> T
            | [T < _ToStr] (T, T, *T method_names) -> Array[T]
            | (interned, interned, *interned method_names) -> Array[interned]
            | [T < _ToAry[interned]] (T method_names) -> T
```

Reply:
Why must the RHS be interfaces?
… *tests* RubyMine says the RHS can also be Classes and Modules, but not Type Variables, let alone arbitrary types (here, a `nil`able). Wynaut? !featurerequest

Reply:
do it

Reply:
#1471